### PR TITLE
HBW-364 Fix select table in edge cases

### PR DESCRIPTION
**Pull request by sergeichukd**
* Initialisation of rows
* Allowed table states

### CODE REVIEW

File: hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx [hunk 1]
```diff
@@ -25,13 +25,31 @@ modulejs.define('HBWFormSelectTable',
         };
       }
 
-      getInputValues = (value) => {
+      initNullablaTable (value) {
         if (this.props.params.multi) {
           return JSON.parse(value) || [];
         } else {
-          return value || '';
+          return value || 'null';
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx:32]: It looks weird to me. We submit string `'null'` when actually should submit nothing. Also, for the 'multi' mode, we can deselect all values. But in the 'single' mode, we should click on the special row. Not sure we should change it in this PR but it will be a good next step I think.

File: hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx [hunk 2]
```diff
@@ -25,13 +25,31 @@ modulejs.define('HBWFormSelectTable',
...
+
+      initTable (value) {
+        if (this.props.params.nullable) {
+          return this.initNullablaTable(value);
+        } else {
+          return this.initNotNullablaTable(value);
+        }
+      }
+
+      getFirstRowId = () => String(this.getChoices()[0][0])
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx:52]: Why do you need to cast it to string?

File: hbw/app/javascript/packs/hbw/components/shared/hoc/select.js [hunk 3]
```diff
@@ -41,6 +41,7 @@ export default WrappedComponent => class WithSelect extends Component {
   };
 
   isEqual = (a, b) => {
+    const { multi } = this.props.params;
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/shared/hoc/select.js:44]: New line

File: hbw/app/javascript/packs/hbw/components/shared/hoc/select.js [hunk 4]
```diff
@@ -50,6 +51,12 @@ export default WrappedComponent => class WithSelect extends Component {
     if (((a === '') && (b === null)) || ((a === null) && (b === ''))) {
       return true;
     }
+    if (((a === 'null') && (b === null)) || ((a === null) && (b === 'null'))) {
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/shared/hoc/select.js:54]: :(

File: hbw/app/javascript/packs/hbw/components/shared/hoc/select.js [hunk 5]
```diff
@@ -50,6 +51,12 @@ export default WrappedComponent => class WithSelect extends Component {
...
+      return true;
+    }
+    if (multi && (((a?.length === 0) && (b === null)) || ((a === null) && (b?.length === 0)))) {
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/shared/hoc/select.js:57]: Could you provide some case descrption? Can't see why do we need it.

File: spec/hbw/features/select_table_spec.rb [hunk 6]
```diff
@@ -111,25 +185,14 @@
       expect_widget_presence
     end
 
-    scenario 'multi = false' do
```
Comment (TimAle) [inline spec/hbw/features/select_table_spec.rb:114]: Why did you remove it?

File: spec/hbw/features/select_table_spec.rb [hunk 7]
```diff
@@ -59,9 +71,48 @@
...
+
+    scenario 'not empty bp variable, nullale, multi = true' do
+      click_and_wait 'ORD-41'
+
+      real_initial_table_values = {
+        table_options:          options_in_table_with_label('homsOrderDataSelect'),
+        selected_options:       selected_options_in_table_with_label('homsOrderDataSelect'),
+        table_options_multi:    options_in_table_with_label('homsOrderDataSelectMulti'),
+        selected_options_multi: selected_options_in_table_with_label('homsOrderDataSelectMulti')
+      }
```
Comment (TimAle) [inline spec/hbw/features/select_table_spec.rb:106]: Why do you have two tables here?

**Review by TimAle — CHANGES_REQUESTED** [review]

File: hbw/app/javascript/packs/hbw/components/shared/hoc/select.js (hunk 4 above)
Comment (sergeichukd) [reply hbw/app/javascript/packs/hbw/components/shared/hoc/select.js:54]: wtf???

File: hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx [hunk 8]
```diff
@@ -15,7 +15,7 @@ modulejs.define('HBWFormSelectTable',
       constructor (props, context) {
         super(props, context);
 
-        const value = this.getInputValues(props.params.current_value);
+        const value = this.initTable(props.params.current_value);
```
[the remark this answers is not in the document]
Comment (sergeichukd) [reply hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx]: corrected name to `initTableValue`

File: hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx [hunk 9]
```diff
@@ -25,13 +25,31 @@ modulejs.define('HBWFormSelectTable',
         };
       }
 
-      getInputValues = (value) => {
+      initNullablaTable (value) {
```
[the remark this answers is not in the document]
Comment (sergeichukd) [reply hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx]: It's a typo. Thanks, I'll correct it

File: hbw/app/javascript/packs/hbw/components/shared/hoc/select.js (hunk 3 above)
Comment (sergeichukd) [reply hbw/app/javascript/packs/hbw/components/shared/hoc/select.js:44]: done

File: hbw/app/javascript/packs/hbw/components/shared/hoc/select.js (hunk 4 above)
Comment (sergeichukd) [reply hbw/app/javascript/packs/hbw/components/shared/hoc/select.js:55]: There is 'null' value in nullable select tables with multi = false.
    We should parse it like a null
    It will be used in hasValueInChoices function

File: hbw/app/javascript/packs/hbw/components/shared/hoc/select.js (hunk 5 above)
Comment (sergeichukd) [reply hbw/app/javascript/packs/hbw/components/shared/hoc/select.js:58]: In nullable multi = true tables we can deselect all rows so value of the table will be []
    We should parse it like a null value.
    It will be used in hasValueInChoices function

File: spec/hbw/features/select_table_spec.rb (hunk 6 above)
Comment (sergeichukd) [reply spec/hbw/features/select_table_spec.rb:114]: Because functionality of select table is changed and this case is not actual

File: hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx (hunk 1 above)
Comment (sergeichukd) [reply hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx]: Ok, I wil revert it

File: hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx (hunk 2 above)
Comment (sergeichukd) [reply hbw/app/javascript/packs/hbw/components/form/select_table.js.jsx]: Because select element has IDs of type string
I moved translation to  String outside of getFirstRowId function

File: spec/hbw/features/select_table_spec.rb (hunk 7 above)
Comment (sergeichukd) [reply spec/hbw/features/select_table_spec.rb:106]: I've separated checking of two tables to different tests

**Review by TimAle — APPROVED** [review]
